// @tencent/workflow-api/lib/catalog.js
const iconv = require('iconv-lite')
const req = require('./req')
let catalog = module.exports = {}

catalog.create = (name, cname, parent) => {
  return new Promise((resolve, reject) => {
    let url = 'http://wizard2.webdev.com/cgi-bin/catalog/catalog_add'
    let data = {
      returnurl: '/cgi-bin/catalog/catalog_main?short=' + parent,
      hiddenstatus: 0,
      fshort: name,
      fcnname: iconv.encode(cname, 'GBK'),
      feblong: 1,
      fcatalogtype: 1,
      fwpacheck: 0,
      faddwaptitle: 0,
      fisimportant: 0,
      catalog_comment_id: 0,
      fparent: parent
    }

    req.post(url, data)
      .then((res) => {
        let ret = {
          code: 0,
          msg: '创建栏目成功'
        }

        if (res.code === 0) {
          let html = res.data.html

          if (html.indexOf('fail') > -1) {
            let reg = /<strong>(.*)<\/strong>/ig
            let arr = reg.exec(html)

            ret['code'] = 303
            ret['msg'] = arr[1]
          } else {
            ret['data'] = {}
            ret['data']['cname'] = cname
          }
        } else {
          ret['code'] = res['code']
          ret['msg'] = '创建栏目失败'
        }

        resolve(ret)
      }, (err) => {
        reject(err)
      })
  })
}

catalog.query = function () {

}
